File: frontend/src/types/document.ts
import type { Category } from './category'
import type { Department, DepartmentSummary } from './department'
import type { DocumentVersion } from './documentVersion'

export type ClassificationLevel = 'PUBLIC' | 'INTERNAL' | 'CONFIDENTIAL' | 'SECRET'

export type DocumentStatus =
  | 'DRAFT'
  | 'SUBMITTED'
  | 'UNDER_REVIEW'
  | 'APPROVED'
  | 'REJECTED'
  | 'ARCHIVED'

export interface Document {
  documentId: string
  referenceNumber: string | null
  title: string
  description: string | null
  category: Category
  department: Department | null
  otherDepartmentName: string | null
  classification: ClassificationLevel
  status: DocumentStatus
  currentVersion: DocumentVersion | null
  archivedAt: string | null
  createdAt: string
  updatedAt: string
  uploadedBy: {
    username: string
    fullName: string
    email: string
    department: DepartmentSummary | null
  } | null
  createdBy: string | null
  updatedBy: string | null
}

export interface CreateDocumentInput {
  title: string
  description?: string
  categoryId: string
  departmentId?: string
  otherDepartmentName?: string
  classification?: ClassificationLevel
}

export interface UpdateDocumentInput {
  title?: string
  description?: string
  categoryId?: string
  departmentId?: string
  otherDepartmentName?: string
  classification?: ClassificationLevel
}

File: backend/src/main/java/com/digitalarchive/domain/entity/Document.java
package com.digitalarchive.domain.entity;

import com.digitalarchive.domain.enums.ClassificationLevel;
import com.digitalarchive.domain.enums.DocumentStatus;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "documents")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Document {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "document_id", updatable = false, nullable = false)
    private UUID documentId;

    // Populated by a trigger (see V6__create_triggers.sql) — never set from Java
    @Column(name = "reference_number", insertable = false, updatable = false)
    private String referenceNumber;

    @Column(name = "title", nullable = false, length = 300)
    private String title;

    @Column(name = "description", columnDefinition = "text")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id")
    private Department department;

    @Column(name = "other_department_name", length = 150)
    private String otherDepartmentName;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "classification", nullable = false)
    private ClassificationLevel classification;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "status", nullable = false)
    private DocumentStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "current_version_id")
    private DocumentVersion currentVersion;

    @Column(name = "archived_at")
    private OffsetDateTime archivedAt;

    @Column(name = "created_at", insertable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", insertable = false, updatable = false)
    private OffsetDateTime updatedAt;

    @Column(name = "created_by", nullable = false)
    private UUID createdBy;

    @Column(name = "uploader_username", nullable = false, length = 150)
    private String uploaderUsername;

    @Column(name = "uploader_full_name", nullable = false, length = 200)
    private String uploaderFullName;

    @Column(name = "uploader_email", nullable = false, length = 255)
    private String uploaderEmail;

    @Column(name = "uploader_department_id")
    private UUID uploaderDepartmentId;

    @Column(name = "uploader_department_name", length = 150)
    private String uploaderDepartmentName;

    @Column(name = "updated_by")
    private UUID updatedBy;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    @Column(name = "deleted_by")
    private UUID deletedBy;
}
